SerialPrograms/Source/PokemonSV/Inference/ItemPrinter/PokemonSV_ItemPrinterMaterialDetector.cpp

```diff
@@ -0,0 +1,289 @@
+            {0xffb0b0b0, 0xffffffff, true},
+            {0xffc0c0c0, 0xffffffff, true},
+            {0xffd0d0d0, 0xffffffff, true},
+            {0xffe0e0e0, 0xffffffff, true},
+            {0xfff0f0f0, 0xffffffff, true},
+        };
+    }else{
+        filters = 
+        {
+            // black text filters
+            // {0xff000000, 0xff101010, true},
+            // {0xff000000, 0xff202020, true},
+            {0xff000000, 0xff303030, true},
+            {0xff000000, 0xff404040, true},
+            {0xff000000, 0xff505050, true},
+            {0xff000000, 0xff606060, true},
+            {0xff000000, 0xff707070, true},
+            {0xff000000, 0xff808080, true},
+        };
+    }
+
+    std::vector<std::pair<ImageRGB32, size_t>> filtered = to_blackwhite_rgb32_range(
+        cropped,
+        filters
+    );
+
+    SpinLock lock;
+    std::map<int16_t, int8_t> candidates;
+    std::vector<std::unique_ptr<AsyncTask>> tasks(filtered.size());
+    // for (size_t c = 0; c < filtered.size(); c++){
+    //     filtered[c].first.save("DebugDumps/test-" + std::to_string(c) + ".png");
+    // }
+    for (size_t c = 0; c < filtered.size(); c++){
+        tasks[c] = dispatcher.dispatch([&, c]{
+            int num = OCR::read_number(logger, filtered[c].first);
+            WriteSpinLock lg(lock);
+            candidates[(int16_t)num]++;
+        });
+    }
+
+    //  Wait for everything.
+    for (auto& task : tasks){
+        task->wait_and_rethrow_exceptions();
+    }
+
+    std::pair<int16_t, int8_t> best;
+    std::pair<int16_t, int8_t> second_best;
+    for (const auto& item : candidates){
+        logger.log("Candidate OCR: " + std::to_string(item.first) + "; x" + std::to_string(item.second));
+        if (item.second >= best.second){
+            second_best = best;
+            best = item;
+        }
+        // std::cout << "Best: " << std::to_string(best.first) << "; x" << std::to_string(best.second) << std::endl;
+        // std::cout << "Second Best: " << std::to_string(second_best.first) << "; x" << std::to_string(second_best.second) << std::endl;
+    }
+
+    if (best.second > second_best.second + 3){
+        return best.first;
+    }else{
+        logger.log("Ambiguous or multiple results with normal number OCR. Use Waterfill number OCR.");
+        if(is_white_text){
+            return (int16_t)OCR::read_number_waterfill(logger, cropped, 0xff808080, 0xffffffff);
+        }else{
+            return (int16_t)OCR::read_number_waterfill(logger, cropped, 0xff000000, 0xff808080);
+        }
+    }
+
+}
+
+// return row number where Happiny dust is located on screen
+// keep pressing DPAD_RIGHT until Happiny dust is on screen
+// check each row on the screen for Happiny Dust
+int8_t ItemPrinterMaterialDetector::find_happiny_dust_row_index(
+    AsyncDispatcher& dispatcher,
+    ConsoleHandle& console, BotBaseContext& context
+) const{
+    int8_t value_68_row_index;
+    for (size_t c = 0; c < 10; c++){
```

Comment:
10 is the max # of times to fully page down?

Reply:
No. There are around 181 materials. So should take 18-19 times to fully page down.

But Happiny Dust is number 73, so should take 8 times to page down. As far as I know, the items are always in the same order. 

I'm fine with increasing the max number of times to scroll, just to account for button drops, or in case the material is in a different order than expected.